ext/opcache/shared_alloc_mmap.c

```diff
@@ -40,6 +40,105 @@
 # define MAP_HUGETLB MAP_ALIGNED_SUPER
 #endif
 
+#if defined(__x86_64__) && defined(__linux__)
+/* On Linux OS, this function returns the start address of the first PHP segment
+   through parsing /proc/self/maps file; on any other OSes, simply returns NULL.
+*/
+static void* get_php_start_addr(void)
+{
+	/* Used to parse each field for the line in /proc/self/maps file */
+	long unsigned int start, end, offset, inode;
+	char perm[5], dev[10], name[MAXPATHLEN];
+	int ret;
+	FILE *f = NULL;
+
+	f = fopen("/proc/self/maps", "r");
+	if (!f) {
+		return NULL;
+	}
+
+	/* Only get the start address of the first PHP segment */
+	ret = fscanf(f, "%lx-%lx %4s %lx %9s %ld %s\n",
+					&start, &end, perm, &offset, dev, &inode, name);
+	fclose(f); f=NULL;
+
+	/* Expect to get seven fields */
+	if (7 == ret) {
+		return (void*)start;
+	}
+	return NULL;
+}
+
+/*
+	This function tries to create segment starting from address prior to
+	the first segment of PHP process and returns the newly-allocated address,
+	or MAP_FAILED for failure.
```

Comment:
Why only directly prior? Why not after?
I think, it's better to analyse segments laying prior and after PHP code segments, find nearest free and big enough segments and then use them for SHM allocation.

Reply:
@dstogov Thanks. That's exactly what we are trying to do programmatically in our new patch.  pls stay tuned.
We manually checked maps file and found that there is usually big space before PHP .text segment. Immediately after PHP .text segment, we often see data segment and heap segment. So it is likely we still hit candidate space before rather than after PHP .text in program.